### PR TITLE
[SYCL][CUDA] Add missing barrier to collectives

### PR DESCRIPTION
**Pull request by Pennycook**
SYCL sub-group and group functions should act as synchronization points.
Group collectives need a barrier at the end to ensure that back-to-back
collectives do not lead to a race condition.

Note that the barrier at the beginning of each collective occurs after
each work-item writes its partial results to the scratch space.  This is
assumed safe because only the collective functions can access the space,
and collective functions must be encountered in uniform control flow; any
work-item encountering a collective function can assume it is safe to use
the scratch space, because all work-items in the same work-group must have
either executed no collective functions or the barrier at the end of a previous
collective function.

Signed-off-by: John Pennycook <john.pennycook@intel.com>

### CODE REVIEW

**Review by anton-v-gorshkov — DISMISSED** [review]
Looks fine for me.

**Review by bader — APPROVED** [review]
Is there a regression test covering this code?

**Comment by Pennycook** [comment]
> Is there a regression test covering this code?

There isn't, and I've struggled to write one.  Because it's a race condition, the problem wouldn't show up on every run of the test.  I'm also having trouble reproducing the failure at all with a simple test-case (with two back-to-back reductions) on the NVIDIA GPU that I use for testing, suggesting it might be quite hard to trigger in practice.

**Comment by bader** [comment]
> > Is there a regression test covering this code?
> 
> There isn't, and I've struggled to write one. Because it's a race condition, the problem wouldn't show up on every run of the test. I'm also having trouble reproducing the failure at all with a simple test-case (with two back-to-back reductions) on the NVIDIA GPU that I use for testing, suggesting it might be quite hard to trigger in practice.

I realize the it reproducer might be not 100% reliable, but a simple test-case sounds better than nothing. I see value in adding such test case as I suppose there is a chance it should be able to catch issues in collectives considering that we run regression tests multiple times. In addition to that it will catch issues not related to data races.
What do you think about it?

**Comment by Pennycook** [comment]
> What do you think about it?

Makes sense.  I've added a simple regression test in 14438de.

File: sycl/test/regression/back_to_back_collectives.cpp [hunk 1]
```diff
@@ -0,0 +1,64 @@
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -o %t.out
+// RUN: %HOST_RUN_PLACEHOLDER %t.out
+// RUN: %CPU_RUN_PLACEHOLDER %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
```
Comment (bader) [inline sycl/test/regression/back_to_back_collectives.cpp:4]: This should be located inside [sycl/test/on-device](https://github.com/intel/llvm/tree/sycl/sycl/test/on-device) directory according to the guidelines from [Get Started Guide](https://github.com/intel/llvm/blob/sycl/sycl/doc/GetStartedGuide.md#test-dpc-toolchain).

**Review by bader — DISMISSED** [review]
LGTM except one minor comment.
Thanks!

File: sycl/test/regression/back_to_back_collectives.cpp (hunk 1 above)
Comment (Pennycook) [reply sycl/test/regression/back_to_back_collectives.cpp]: Oops.  I'll go and read the new guidelines.

**Review by bader — APPROVED** [review]

**Review by alexbatashev — DISMISSED** [review]

File: sycl/test/regression/back_to_back_collectives.cpp (hunk 1 above)
Comment (romanovvlad) [reply sycl/test/regression/back_to_back_collectives.cpp]: I think the test should be even moved to https://github.com/intel/llvm-test-suite

File: sycl/test/on-device/back_to_back_collectives.cpp [hunk 2]
```diff
@@ -0,0 +1,64 @@
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -o %t.out
```
Comment (bader) [inline sycl/test/on-device/back_to_back_collectives.cpp:1]:
```suggestion
// RUN: %clangxx -fsycl -fsycl-unnamed-lambda -fsycl-targets=%sycl_triple %s -o %t.out
```

File: sycl/test/on-device/back_to_back_collectives.cpp [hunk 3]
```diff
@@ -0,0 +1,64 @@
+// RUN: %clangxx -fsycl -fsycl-unnamed-lambda -fsycl-targets=%sycl_triple %s -o %t.out
+// RUN: %HOST_RUN_PLACEHOLDER %t.out
```
Comment (bader) [inline sycl/test/on-device/back_to_back_collectives.cpp:2]:
```suggestion
// RUN: %RUN_ON_HOST %t.out
```

File: sycl/test/on-device/back_to_back_collectives.cpp [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+int main() {
+  queue q;
+  if (q.get_device().is_host()) {
+    std::cout << "Skipping test\n";
+    return 0;
+  }
+
+  // Use max work-group size to maximize chance of race
+  int N = q.get_device().get_info<info::device::max_work_group_size>();
```
Comment (bader) [inline sycl/test/on-device/back_to_back_collectives.cpp:21]: Probably it's not a best choice. This test is failing on CPU device with "out of resources" error.

Comment (Pennycook) [reply sycl/test/on-device/back_to_back_collectives.cpp]: You're right.  I forgot that calling the work-group functions might impact the maximum work-group size supported by the kernel.  I've made the query kernel-specific in https://github.com/intel/llvm/pull/2990/commits/6d9918b07eed9b58749014d6af16f8b41c86392e.

**Review by alexbatashev — APPROVED** [review]

**Review by bader — APPROVED** [review]

**Pull request merged**